IntelliJ is very instable, freeze, high CPU, auto-complete lag, IDE crash - 1.15.0.308

Hello,

Since the last version of IntelliJ (2022.3 UE)
Build #IU-223.7571.182, built on November 29, 2022 with Quarkus Tools (1.15.0.308).

IntelliJ is very instable, freeze, high CPU, auto-complete lag, IDE crash :

**The stack trace :**
Internal error: com.intellij.openapi.project.IndexNotReadyException: Please change caller according to com.intellij.openapi.project.IndexNotReadyException documentation

java.util.concurrent.CompletionException: com.intellij.openapi.project.IndexNotReadyException: Please change caller according to com.intellij.openapi.project.IndexNotReadyException documentation
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at com.redhat.devtools.intellij.quarkus.lsp.QuarkusLanguageClient.lambda$runAsBackground$2(QuarkusLanguageClient.java:124)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:190)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:591)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:666)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:622)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:590)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:177)
	at com.redhat.devtools.intellij.quarkus.lsp.QuarkusLanguageClient.lambda$runAsBackground$3(QuarkusLanguageClient.java:120)
	at com.redhat.devtools.intellij.quarkus.lsp.QuarkusLanguageClient.lambda$runAsBackground$4(QuarkusLanguageClient.java:130)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: com.intellij.openapi.project.IndexNotReadyException: Please change caller according to com.intellij.openapi.project.IndexNotReadyException documentation
	at com.intellij.openapi.project.IndexNotReadyException.create(IndexNotReadyException.java:66)
	at com.intellij.util.indexing.FileBasedIndexImpl.handleDumbMode(FileBasedIndexImpl.java:841)
	at com.intellij.util.indexing.FileBasedIndexImpl.ensureUpToDate(FileBasedIndexImpl.java:789)
	at com.intellij.psi.stubs.StubIndexEx.getContainingIds(StubIndexEx.java:390)
	at com.intellij.psi.stubs.StubIndexEx.processElements(StubIndexEx.java:167)
	at com.intellij.psi.stubs.StubIndex.getElements(StubIndex.java:103)
	at com.intellij.psi.stubs.StubIndex.getElements(StubIndex.java:91)
	at com.intellij.psi.impl.java.stubs.index.JavaFullClassNameIndex.get(JavaFullClassNameIndex.java:34)
	at com.intellij.psi.impl.file.impl.JavaFileManagerImpl.doFindClasses(JavaFileManagerImpl.java:80)
	at com.intellij.psi.impl.file.impl.JavaFileManagerImpl.findClass(JavaFileManagerImpl.java:103)
	at com.intellij.psi.impl.PsiElementFinderImpl.findClass(PsiElementFinderImpl.java:40)
	at com.intellij.psi.impl.JavaPsiFacadeImpl.doFindClass(JavaPsiFacadeImpl.java:93)
	at com.intellij.psi.impl.JavaPsiFacadeImpl.findClass(JavaPsiFacadeImpl.java:71)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.utils.PsiTypeUtils.findType(PsiTypeUtils.java:145)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.jaxrs.java.JaxRsCodeLensParticipant.isAdaptedForCodeLens(JaxRsCodeLensParticipant.java:65)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.PropertiesManagerForJava.lambda$collectCodeLens$2(PropertiesManagerForJava.java:145)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:178)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.PropertiesManagerForJava.collectCodeLens(PropertiesManagerForJava.java:146)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.PropertiesManagerForJava.lambda$codeLens$1(PropertiesManagerForJava.java:130)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:860)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.PropertiesManagerForJava.codeLens(PropertiesManagerForJava.java:123)
	at com.redhat.devtools.intellij.quarkus.lsp.QuarkusLanguageClient.lambda$getJavaCodelens$12(QuarkusLanguageClient.java:173)
	at com.redhat.devtools.intellij.quarkus.lsp.QuarkusLanguageClient.lambda$runAsBackground$2(QuarkusLanguageClient.java:122)
	... 16 more

--- a/build.gradle
+++ b/build.gradle
@@ -1,6 +1,6 @@
 
 plugins {
-    id "org.jetbrains.intellij" version "1.9.0"
+    id "org.jetbrains.intellij" version "1.10.1"
     id 'idea'
     id 'java'
     id 'org.jetbrains.kotlin.jvm' version '1.6.0'

--- a/src/main/java/com/redhat/devtools/intellij/quarkus/lsp/QuarkusLanguageClient.java
+++ b/src/main/java/com/redhat/devtools/intellij/quarkus/lsp/QuarkusLanguageClient.java
@@ -11,9 +11,6 @@
 package com.redhat.devtools.intellij.quarkus.lsp;
 
 import com.intellij.openapi.module.Module;
-import com.intellij.openapi.progress.EmptyProgressIndicator;
-import com.intellij.openapi.progress.ProgressManager;
-import com.intellij.openapi.project.DumbService;
 import com.intellij.openapi.project.Project;
 import com.intellij.openapi.roots.libraries.Library;
 import com.intellij.openapi.vfs.VirtualFile;
@@ -21,13 +18,11 @@ import com.intellij.util.messages.MessageBusConnection;
 import com.redhat.devtools.intellij.lsp4mp4ij.psi.core.ProjectLabelManager;
 import com.redhat.devtools.intellij.lsp4mp4ij.psi.core.PropertiesManager;
 import com.redhat.devtools.intellij.lsp4mp4ij.psi.core.PropertiesManagerForJava;
-import com.redhat.devtools.intellij.lsp4mp4ij.psi.core.project.IConfigSource;
-import com.redhat.devtools.intellij.lsp4mp4ij.psi.core.project.IConfigSourceProvider;
 import com.redhat.devtools.intellij.lsp4mp4ij.psi.core.project.PsiMicroProfileProjectManager;
 import com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.ls.PsiUtilsLSImpl;
 import com.redhat.devtools.intellij.quarkus.QuarkusModuleUtil;
 import com.redhat.devtools.intellij.quarkus.QuarkusProjectService;
-import com.redhat.devtools.intellij.quarkus.lsp4ij.LanguageClientImpl;
+import com.redhat.devtools.intellij.quarkus.lsp4ij.IndexAwareLanguageClient;
 import org.apache.commons.lang3.tuple.Pair;
 import org.eclipse.lsp4j.CodeAction;
 import org.eclipse.lsp4j.CodeLens;
@@ -58,11 +53,10 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
-import java.util.function.Supplier;
 import java.util.stream.Collectors;
 
 
-public class QuarkusLanguageClient extends LanguageClientImpl implements MicroProfileLanguageClientAPI, QuarkusProjectService.Listener {
+public class QuarkusLanguageClient extends IndexAwareLanguageClient implements MicroProfileLanguageClientAPI, QuarkusProjectService.Listener {
   private static final Logger LOGGER = LoggerFactory.getLogger(QuarkusLanguageClient.class);
   private static final String JAVA_FILE_EXTENSION = "java";
 
@@ -113,68 +107,48 @@ public class QuarkusLanguageClient extends LanguageClientImpl implements MicroPr
     return PsiMicroProfileProjectManager.getInstance(project.getProject()).isConfigSource(file);
   }
 
-
-  <R> CompletableFuture<R> runAsBackground(String title, Supplier<R> supplier) {
-    CompletableFuture<R> future = new CompletableFuture<>();
-    CompletableFuture.runAsync(() -> {
-      Runnable task = () -> ProgressManager.getInstance().runProcess(() -> {
-        try {
-          future.complete(supplier.get());
-        } catch (Throwable t) {
-          future.completeExceptionally(t);
-        }
-      }, new EmptyProgressIndicator());
-      if (DumbService.getInstance(getProject()).isDumb()) {
-        DumbService.getInstance(getProject()).runWhenSmart(task);
-      } else {
-        task.run();
-      }
-    });
-    return future;
-  }
-
   @Override
   public CompletableFuture<MicroProfileProjectInfo> getProjectInfo(MicroProfileProjectInfoParams params) {
-    return runAsBackground("Computing project information", () -> PropertiesManager.getInstance().getMicroProfileProjectInfo(params, PsiUtilsLSImpl.getInstance(getProject())));
+    return runAsBackground("Computing project information", monitor -> PropertiesManager.getInstance().getMicroProfileProjectInfo(params, PsiUtilsLSImpl.getInstance(getProject())));
   }
 
   @Override
   public CompletableFuture<Hover> getJavaHover(MicroProfileJavaHoverParams javaParams) {
-    return runAsBackground("Computing Java hover", () -> PropertiesManagerForJava.getInstance().hover(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
+    return runAsBackground("Computing Java hover", monitor -> PropertiesManagerForJava.getInstance().hover(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
   }
 
   @Override
   public CompletableFuture<List<PublishDiagnosticsParams>> getJavaDiagnostics(MicroProfileJavaDiagnosticsParams javaParams) {
-    return runAsBackground("Computing Java diagnostics", () -> PropertiesManagerForJava.getInstance().diagnostics(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
+    return runAsBackground("Computing Java diagnostics", monitor -> PropertiesManagerForJava.getInstance().diagnostics(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
   }
 
   @Override
   public CompletableFuture<Location> getPropertyDefinition(MicroProfilePropertyDefinitionParams params) {
-    return runAsBackground("Computing property definition", () -> PropertiesManager.getInstance().findPropertyLocation(params, PsiUtilsLSImpl.getInstance(getProject())));
+    return runAsBackground("Computing property definition", monitor -> PropertiesManager.getInstance().findPropertyLocation(params, PsiUtilsLSImpl.getInstance(getProject())));
   }
 
   @Override
   public CompletableFuture<ProjectLabelInfoEntry> getJavaProjectlabels(MicroProfileJavaProjectLabelsParams javaParams) {
-    return runAsBackground("Computing Java projects labels", () -> ProjectLabelManager.getInstance().getProjectLabelInfo(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
+    return runAsBackground("Computing Java projects labels", monitor -> ProjectLabelManager.getInstance().getProjectLabelInfo(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
   }
 
   @Override
   public CompletableFuture<JavaFileInfo> getJavaFileInfo(MicroProfileJavaFileInfoParams javaParams) {
-    return runAsBackground("Computing Java file info", () -> PropertiesManagerForJava.getInstance().fileInfo(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
+    return runAsBackground("Computing Java file info", monitor -> PropertiesManagerForJava.getInstance().fileInfo(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
   }
 
   @Override
   public CompletableFuture<CompletionList> getJavaCompletion(MicroProfileJavaCompletionParams javaParams) {
-    return runAsBackground("Computing Java completion", () -> PropertiesManagerForJava.getInstance().completion(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
+    return runAsBackground("Computing Java completion", monitor -> PropertiesManagerForJava.getInstance().completion(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
   }
 
   @Override
   public CompletableFuture<List<? extends CodeLens>> getJavaCodelens(MicroProfileJavaCodeLensParams javaParams) {
-    return runAsBackground("Computing Java codelens", () -> PropertiesManagerForJava.getInstance().codeLens(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
+    return runAsBackground("Computing Java codelens", monitor -> PropertiesManagerForJava.getInstance().codeLens(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
   }
 
   @Override
   public CompletableFuture<List<CodeAction>> getJavaCodeAction(MicroProfileJavaCodeActionParams javaParams) {
-    return runAsBackground("Computing Java code actions", () -> (List<CodeAction>) PropertiesManagerForJava.getInstance().codeAction(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
+    return runAsBackground("Computing Java code actions", monitor -> (List<CodeAction>) PropertiesManagerForJava.getInstance().codeAction(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
   }
 }

--- a/src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/IndexAwareLanguageClient.java
+++ b/src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/IndexAwareLanguageClient.java
@@ -1,0 +1,109 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at https://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package com.redhat.devtools.intellij.quarkus.lsp4ij;
+
+import com.intellij.openapi.application.ReadAction;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.DumbService;
+import com.intellij.openapi.project.IndexNotReadyException;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.libraries.Library;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.util.messages.MessageBusConnection;
+import com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.ls.PsiUtilsLSImpl;
+import com.redhat.devtools.intellij.quarkus.QuarkusModuleUtil;
+import com.redhat.devtools.intellij.quarkus.QuarkusProjectService;
+import com.redhat.devtools.intellij.qute.psi.QuteSupportForJava;
+import com.redhat.devtools.intellij.qute.psi.QuteSupportForTemplate;
+import com.redhat.qute.commons.JavaTypeInfo;
+import com.redhat.qute.commons.ProjectInfo;
+import com.redhat.qute.commons.QuteJavaCodeLensParams;
+import com.redhat.qute.commons.QuteJavaDefinitionParams;
+import com.redhat.qute.commons.QuteJavaDiagnosticsParams;
+import com.redhat.qute.commons.QuteJavaDocumentLinkParams;
+import com.redhat.qute.commons.QuteJavaTypesParams;
+import com.redhat.qute.commons.QuteProjectParams;
+import com.redhat.qute.commons.QuteResolvedJavaTypeParams;
+import com.redhat.qute.commons.ResolvedJavaTypeInfo;
+import com.redhat.qute.commons.datamodel.DataModelParameter;
+import com.redhat.qute.commons.datamodel.DataModelProject;
+import com.redhat.qute.commons.datamodel.DataModelTemplate;
+import com.redhat.qute.commons.datamodel.JavaDataModelChangeEvent;
+import com.redhat.qute.commons.datamodel.QuteDataModelProjectParams;
+import com.redhat.qute.commons.usertags.QuteUserTagParams;
+import com.redhat.qute.commons.usertags.UserTagInfo;
+import com.redhat.qute.ls.api.QuteLanguageClientAPI;
+import com.redhat.qute.ls.api.QuteLanguageServerAPI;
+import org.apache.commons.lang3.tuple.Pair;
+import org.eclipse.lsp4j.CodeLens;
+import org.eclipse.lsp4j.DocumentLink;
+import org.eclipse.lsp4j.Location;
+import org.eclipse.lsp4j.PublishDiagnosticsParams;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class IndexAwareLanguageClient extends LanguageClientImpl {
+  private static final Logger LOGGER = LoggerFactory.getLogger(IndexAwareLanguageClient.class);
+
+  public IndexAwareLanguageClient(Project project) {
+    super(project);
+  }
+
+  protected <R> CompletableFuture<R> runAsBackground(String title, Function<ProgressIndicator, R> function) {
+    CompletableFuture<R> future = new CompletableFuture<>();
+    CompletableFuture.runAsync(() -> {
+      Runnable task = () -> ProgressManager.getInstance().run(new Task.Backgroundable(getProject(), title) {
+        @Override
+        public void run(@NotNull ProgressIndicator indicator) {
+          runTask(indicator, future, function);
+        }
+
+        @Override
+        public boolean isHeadless() {
+          return true;
+        }
+      });
+      if (DumbService.getInstance(getProject()).isDumb()) {
+        DumbService.getInstance(getProject()).runWhenSmart(task);
+      } else {
+        task.run();
+      }
+    });
+    return future;
+  }
+
+  private <R> void runTask(@NotNull ProgressIndicator indicator, CompletableFuture<R> future, Function<ProgressIndicator, R> function) {
+    boolean done = false;
+    for(int i=0; !done && i < 10;++i) {
+      try {
+        future.complete(function.apply(indicator));
+        done = true;
+      } catch (IndexNotReadyException e) {
+      } catch (Throwable t) {
+        future.completeExceptionally(t);
+        done = true;
+      }
+    }
+    if (!done) {
+      DumbService.getInstance(getProject()).runWhenSmart(() -> runTask(indicator, future, function));
+    }
+  }
+}

--- a/src/main/java/com/redhat/devtools/intellij/qute/lsp/QuteLanguageClient.java
+++ b/src/main/java/com/redhat/devtools/intellij/qute/lsp/QuteLanguageClient.java
@@ -12,11 +12,6 @@ package com.redhat.devtools.intellij.qute.lsp;
 
 import com.intellij.openapi.application.ReadAction;
 import com.intellij.openapi.module.Module;
-import com.intellij.openapi.progress.ProgressIndicator;
-import com.intellij.openapi.progress.ProgressManager;
-import com.intellij.openapi.progress.Task;
-import com.intellij.openapi.project.DumbService;
-import com.intellij.openapi.project.IndexNotReadyException;
 import com.intellij.openapi.project.Project;
 import com.intellij.openapi.roots.libraries.Library;
 import com.intellij.openapi.vfs.VirtualFile;
@@ -24,7 +19,7 @@ import com.intellij.util.messages.MessageBusConnection;
 import com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.ls.PsiUtilsLSImpl;
 import com.redhat.devtools.intellij.quarkus.QuarkusModuleUtil;
 import com.redhat.devtools.intellij.quarkus.QuarkusProjectService;
-import com.redhat.devtools.intellij.quarkus.lsp4ij.LanguageClientImpl;
+import com.redhat.devtools.intellij.quarkus.lsp4ij.IndexAwareLanguageClient;
 import com.redhat.devtools.intellij.qute.psi.QuteSupportForJava;
 import com.redhat.devtools.intellij.qute.psi.QuteSupportForTemplate;
 import com.redhat.qute.commons.JavaTypeInfo;
@@ -51,17 +46,15 @@ import org.eclipse.lsp4j.CodeLens;
 import org.eclipse.lsp4j.DocumentLink;
 import org.eclipse.lsp4j.Location;
 import org.eclipse.lsp4j.PublishDiagnosticsParams;
-import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.List;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
-import java.util.function.Function;
 import java.util.stream.Collectors;
 
-public class QuteLanguageClient extends LanguageClientImpl implements QuteLanguageClientAPI, QuarkusProjectService.Listener {
+public class QuteLanguageClient extends IndexAwareLanguageClient implements QuteLanguageClientAPI, QuarkusProjectService.Listener {
   private static final Logger LOGGER = LoggerFactory.getLogger(QuteLanguageClient.class);
 
   private final MessageBusConnection connection;
@@ -94,46 +87,6 @@ public class QuteLanguageClient extends LanguageClientImpl implements QuteLangua
             collect(Collectors.toSet());
     if (!uris.isEmpty()) {
       sendPropertiesChangeEvent(uris);
-    }
-  }
-
-  <R> CompletableFuture<R> runAsBackground(String title, Function<ProgressIndicator, R> function) {
-    CompletableFuture<R> future = new CompletableFuture<>();
-    CompletableFuture.runAsync(() -> {
-      Runnable task = () -> ProgressManager.getInstance().run(new Task.Backgroundable(getProject(), title) {
-        @Override
-        public void run(@NotNull ProgressIndicator indicator) {
-          runTask(indicator, future, function);
-        }
-
-        @Override
-        public boolean isHeadless() {
-          return true;
-        }
-      });
-      if (DumbService.getInstance(getProject()).isDumb()) {
-        DumbService.getInstance(getProject()).runWhenSmart(task);
-      } else {
-        task.run();
-      }
-    });
-    return future;
-  }
-
-  private <R> void runTask(@NotNull ProgressIndicator indicator, CompletableFuture<R> future, Function<ProgressIndicator, R> function) {
-    boolean done = false;
-    for(int i=0; !done && i < 10;++i) {
-      try {
-        future.complete(function.apply(indicator));
-        done = true;
-      } catch (IndexNotReadyException e) {
-      } catch (Throwable t) {
-        future.completeExceptionally(t);
-        done = true;
-      }
-    }
-    if (!done) {
-      DumbService.getInstance(getProject()).runWhenSmart(() -> runTask(indicator, future, function));
     }
   }
 

